### DISCUSSION TITLE
How annotate function return type with an inner class?

### DISCUSSION THREAD

**Discussion by jeffwji** [slick/slick#2421] [Questions] [answered]
I'm using Slick as the database framework, and I want to use the DatabaseConfig to mask the underlying database type, for which I can only seem to define the database schema in the function, so I get the following object:

```
import slick.basic.DatabaseConfig
import slick.jdbc.JdbcProfile
import slick.lifted
import slick.lifted.ProvenShape

import java.util.UUID

object CityTable {
  def apply()(implicit dc: DatabaseConfig[JdbcProfile]) = {
    import dc.profile.api._
    class CityTableSchema(tag: Tag) extends Table[City](tag, "city") {
      def id: Rep[UUID] = column[UUID]("id", O.PrimaryKey)    // This is the primary key column
      def name: Rep[String] = column[String]("name")
      def stateCode: Rep[String] = column[String]("stateCode")
      def * : ProvenShape[City] = (
        id.?,
        name.?,
        stateCode.?
      ) <> (City.tupled, City.unapply)
    }
    lifted.TableQuery[CityTableSchema]
  }
}
```

Now the question is the class `CityTableSchema` is defined in `apply` function. What if I want to refer its type from elsewhere, such as annotating it as the return type of `apply`? The following statement doesn't work, because `CityTableSchema` can't be resolved out of the function.

```
val cityTable: TableQuery[CityTableSchema] = CityTable()
```

**Comment by nafg** [comment, marked as answer]
A function with a persistent scope is a class ;)

```scala
class CityTable(implicit dc: DatabaseConfig[JdbcProfile]) {
    import dc.profile.api._
    class CityTableSchema(tag: Tag) extends Table[City](tag, "city") {
      def id: Rep[UUID] = column[UUID]("id", O.PrimaryKey)    // This is the primary key column
      def name: Rep[String] = column[String]("name")
      def stateCode: Rep[String] = column[String]("stateCode")
      def * : ProvenShape[City] = (
        id.?,
        name.?,
        stateCode.?
      ) <> (City.tupled, City.unapply)
    }
    lifted.TableQuery[CityTableSchema]
}
```

You could also use a trait.

The "elsewhere" code would of course need to provide an implicit `DatabaseConfig[JdbcProfile]`. It could then instantiate the class and import its member, or be in code mixed in with the trait and access its members that way.

That said, you should think if you really need to abstract over databases. Usually you don't want to do that anyway (especially nowadays when you can easily spin up a test database with https://testcontainers.org/ or docker directly) and it will make the code easier to work with. But if you really have a need to abstract over databases, Slick lets you do it and that is one way to do it.
